### PR TITLE
feat: Add --certificate-revocation-list to apiserver

### PR DESCRIPTION
**Pull request by remram44**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Adds optional support for a certificate revocation list (CRL) to the apiserver.

#### Which issue(s) this PR fixes:

See #18982

#### Special notes for your reviewer:

This is similar to how etcd handles CRL files: https://github.com/etcd-io/etcd/pull/8124

#### Does this PR introduce a user-facing change?

```release-note
Added a `--certificate-revocation-list` command line argument to the kube-apiserver. Specifying a value allows you to
configure the API server to reject connections from clients that present revoked certificates.
```

### CODE REVIEW

**Comment by remram44** [comment]
The lint recommends replacing `ParseCRL()` with `ParseRevocationList()`, however that new one doesn't handle PEM-formatted files.

I could inline that part of `ParseCRL()` but is that really desirable?

```go
if bytes.HasPrefix(crlBytes, []byte("-----BEGIN X509 CRL")) {
	block, _ := pem.Decode(crlBytes)
	if block != nil && block.Type == pemType {
		crlBytes = block.Bytes
	}
}
...
```

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go [hunk 1]
```diff
@@ -213,6 +216,10 @@ func (s *SecureServingOptions) AddFlags(fs *pflag.FlagSet) {
 		"If true, SO_REUSEADDR will be used when binding the port. This allows binding "+
 			"to wildcard IPs like 0.0.0.0 and specific IPs in parallel, and it avoids waiting "+
 			"for the kernel to release sockets in TIME_WAIT state. [default=false]")
+
+	fs.StringVar(&s.CertificateRevocationList, "certificate-revocation-list", "", ""+
+		"Certificate revocation list. If provided, clients providing a certificate listed in the "+
```
Comment (sftim) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go:221]:
```suggestion
		"Path to local certificate revocation list. If provided, clients providing a certificate listed in the "+
```

**Comment by sftim** [comment]
Changelog suggestion:
```diff
-Add a new flag --certificate-revocation-list to the apiserver, allowing it to reject connections from clients who present revoked certificates.
+Added a `--certificate-revocation-list` command line argument to the kube-apiserver. Specifying a value allows you to
+configure the API server to reject connections from clients that present revoked certificates.
```

**Comment by sftim** [comment]
> I could inline that part of `ParseCRL()` but is that really desirable?

How about adding this as its own commit, so that people can at least test the change and see everything turn green?

**Comment by sftim** [comment]
:thought_balloon: I don't think we should ship this just for one component.

Each component that can listen for TLS should support the same command line argument and we should ship it all at once. Yes, that means more work for the implementer.

**Comment by enj** [comment]
/hold

This requires a discussion with SIG Auth.  Depending on the outcome of that discussion, we can decide on the next steps (almost certainly a KEP).

**Comment by remram44** [comment]
Sorry for the delay, I had some health issues and had to be in the hospital for a while.

I have applied the suggestions to the flag's help message and the changelog entry, thank you.

> I don't think we should ship this just for one component.

I would argue that this is more important for the apiserver, as it is the external-facing component. AFAIC, this is the only place where I would use such a feature, and should something really bad happen to other certificates, I can easily rotate the entire CA without having to bother users. But I definitely agree with the sentiment, in an ideal world everything that uses certificates should have some strategy to deal with revocation.

**Comment by remram44** [comment]
> How about adding this as its own commit, so that people can at least test the change and see everything turn green?

I inlined the deprecated function, the checks are green

**Comment by remram44** [comment]
Should I rebase or is this a dead-end?

**Comment by enj** [comment]
The [needs KEP](https://docs.google.com/document/d/1sY8fRyRtk4eG9R439z5ao5i9bFuuxilS03XaNlqoni0) doc is tracking the planned work.  There are cert related items on the list, but nothing is currently explicitly tracking revocation.  There are no free cycles in v1.31, but if someone is interested in driving changes in v1.32, then they can add discussion topics to the [agenda](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM).

**Comment by remram44** [comment]
This file is read-only.

Can whatever SIG do whatever it feels is required so we can add support for revoking certificate? Is there a particular reason this is controversial in any way?

**Comment by enj** [comment]
Join the [dev mailing list](https://groups.google.com/a/kubernetes.io/g/dev) to get comment/suggestion access to the docs.

> Can whatever SIG do whatever it feels is required so we can add support for revoking certificate?

If you care about having this feature, you will need to show up to the biweekly meeting and drive the process yourself, including writing the KEP (see the kubernetes/enhancements repo for examples).

> Is there a particular reason this is controversial in any way?

Yes, because revocation lists have unspecified behavior in edge cases such as when they are unreachable, which the KEP would define.  There are also many places in Kube that consume CA data, and we should be consistent in how those work, including revocation support.

This PR itself will not make progress any time soon.  Depending on your needs, a proxy could be used to implement revocation support.

**Comment by remram44** [comment]
How can a local file be "unreachable"?

**Comment by enj** [comment]
> How can a local file be "unreachable"?

I would not be in favor of a local file based approach, as many of the extension points that consume CA data aren't file system based, nor can they handle large raw byte arrays.  Revocation list updates must also be handled.  Thus I expect the CRL to be fetched remotely and refreshed periodically.

**Comment by remram44** [comment]
The apiserver is very much filesystem based though. It gets certificates and keys from the local filesystem, provided by command line arguments. Why should the revocation list use a different mechanism?

My view is that every program that takes in a --client-ca-file should take a --revocation-list. Not having that is a serious security bug.

I am really having trouble understanding the SIG Auth view.

**Comment by enj** [comment]
> The apiserver is very much filesystem based though.

The API server acts as both a client and a server in a variety of scenarios.  It also projects much of that information into the Kube REST API for consumption by other components.  For example, the client CA bundle is reflected in a config map which is consumed by other components, so that they can honor the same client CA bundle.  Those components would be expected to understand the revocation list as well.  CAs already have a `CRLDistributionPoints` field, which would serve as a natural mechanism to consume CRL data from a `http` source.

> My view is that every program that takes in a --client-ca-file should take a --revocation-list.

There are important scenarios outside of just client CA authn to API server.  Client/server scenarios across all components like API server, scheduler, kubelet, etc must be considered, and must be consistent.

> Not having that is a serious security bug.

The rule for CAs is the same as JWT/OIDC authn to the API server.  Neither supports revocation.  Thus all credentials must be short lived.  For JWTs, something like 5 mins is a good approach, and usually an hour for certs.  If you cannot use it in that way, simply do not set the flag.  All of this is optional.

**Comment by remram44** [comment]
You're describing other features with other configuration mechanism and how *those* configuration mechanisms should be extended to support revocation too. That's great!

The only mechanism this PR touches only supports configuration through the command line.

Your point about JWT I don't understand at all.

**Comment by enj** [comment]
> You're describing other features

All of which relate to each other and must be consistent.  Especially when the current flags directly funnel into those features.

> The only mechanism this PR touches only supports configuration through the command line.

We aren't going to expand one place without fixing the others in consistent ways.  This is exactly why you need a KEP to make any progress here.

> Your point about JWT I don't understand at all.

JWTs function the same way as client certs in Kube.  There is no way to revoke them, so they must be short lived.

**Comment by remram44** [comment]
I'm all for consistent, but you are straying from the point that was you not "being in favor of a local file based approach", while the entire client certificate feature of the apiserver uses a local file approach.

Are you trying to change that? I am so confused by what you are arguing for, behind those big words like "consistent", "variety of scenarios", and *other* "extension points".

This one is already file-based and command-line based. There is NO WAY that is changing soon. Why would we wait to have a plan for how revocation works for every single place where anything uses a certificate ever in Kubernetes, before we do the only thing we can possibly do which is provide revocation with the CA *with the same mechanism*?

How would the apiserver even go about loading its revocation list from the apiserver? It can't do that until it accepts connections. There are no alternatives here.

**Comment by enj** [comment]
You seem to have gotten confused somewhere.  Perhaps an [example](https://github.com/kubernetes/kubernetes/compare/master...enj:kubernetes:enj/f/crl_check?w=1) from an old diff may help you understand the approach I would recommend (note that I am not asking you to update your PR in any way - nothing will be merged in regards to cert revocation without a KEP).

> Why would we wait to have a plan for how revocation works for every single place where anything uses a certificate ever in Kubernetes

Because that is how you keep an API driven project consistent over time.  You don't have to build and fix everything at once, but you do have to plan for it.

**Comment by remram44** [comment]
I don't know what to say to that. Your code seems only tangentially related (it's a CLIENT), and the fact that you had to abandon it despite being a SIG member yourself really does not make me want to try and go through the SIG's paperwork process myself.

I can fix the issue for myself, it's too bad for the hundred people that have +1'd that issue over the last 9 years.

**Comment by XANi** [comment]
@enj is there any actual plan on how to handle "user gets their creds stolen" case in K8S somewhere? 

We got to the point where we want to start giving out access to our users in more granular way but I couldn't find anything sensible about it aside from #18982 not happening for last near decade

**Comment by enj** [comment]
@XANi the intention of the cert based auth has always been for testing/break glass scenarios - do not try to use it like an IDP.  If you want to hand out creds to normal users, use the JWT/token webhook integrations via something like [pinniped](https://pinniped.dev/docs/tutorials/supervisor-without-concierge-demo) or [guard](https://github.com/kubeguard/guard) to integrate with an actual IDP.  Pinniped only issues very short lived creds so the lack of revocation for JWTs doesn't matter.  Guard and other token webhooks can choose to implement revocation, though most webooks will also just rely on creds being very short lived.  [Plugins](https://github.com/int128/kubelogin) on the client side make this transparent to the user.  GKE, EKS and AKS all follow this approach.  Mapping users to static, long lived creds should be avoided.

If we ever get to the point of using cluster trust bundles (still alpha today) in scenarios where the API server is the client, I may try to push (again) for it to parse out CRL URLs from the CA bundles.

**Comment by remram44** [comment]
My only concern is kubelets. If someone gets their hand on a drive from a retired node, they have a certificate. This is no "break glass" scenario, in fact it's the only way kubeadm will set up kubelets.

**Comment by neolit123** [comment]
> My only concern is kubelets. If someone gets their hand on a drive from a retired node, they have a certificate. This is no "break glass" scenario, in fact it's the only way kubeadm will set up kubelets.

is a "retired node" also supposed to call "kubeadm reset", because that would clean "/var/lib/kubelet" where the self-signed client certificate of the kubelet would reside?

also, i don't think it's well documented in kubeadm docs but technically you could provision the kubelet.conf kubeconfig (with the client cert) out of band and disable the CSR signing of kubelet client certs and their automatic rotation.
for kubelet serving certs the story is similar.

**Comment by remram44** [comment]
Sure, we could go the "hope we never lose a key" route. Or we could have security, like everybody else does when they use certs..

Having a revocation strategy is really not the niche thing the Kubernetes project seems to think it is.

**Comment by enj** [comment]
> > My only concern is kubelets. If someone gets their hand on a drive from a retired node, they have a certificate. This is no "break glass" scenario, in fact it's the only way kubeadm will set up kubelets.
> 
> is a "retired node" also supposed to call "kubeadm reset", because that would clean "/var/lib/kubelet" where the self-signed client certificate of the kubelet would reside?
> 
> also, i don't think it's well documented in kubeadm docs but technically you could provision the kubelet.conf kubeconfig (with the client cert) out of band and disable the CSR signing of kubelet client certs and their automatic rotation. for kubelet serving certs the story is similar.

Or just lower the max cert lifetime to a day (or whatever interval you are comfortable with) in Kube controller manager and let the kubelet auto rotate and renew its cert.  There is no need to provision long lived creds to kubelets.  The CSR API itself also supports a duration hint field which is honored by all the built-in singers.  That can be used to control cert duration per cert if desired.  If you have an IDP and prefer to manage kubelet identities via it, exec credential providers can be used with the kubelet as well.

**Comment by remram44** [comment]
> just lower the max cert lifetime to a day (or whatever interval you are comfortable with) in Kube controller manager and let the kubelet auto rotate and renew its cert

How would one do this? Renewing the serving certificate is easy, but the kubelet certificate less so. Unless I misunderstands, kubeadm normally renews those?

**Comment by viveksupe** [comment]
@remram44 That seems to be the expected behaviour for `kubeadm`. Agree we need a better revocation solution.

https://github.com/kubernetes/kubernetes/issues/101965#issuecomment-841324607

**Comment by remram44** [comment]
Ok. We can't have the fix in kubelet, because it should be done differently, but the different way is not possible, "by design". Got it.

**Comment by enj** [comment]
For kubelets since `kubeadm` already sets up the right RBAC and such for the kubelet to auto rotate its creds, I think the simplest thing there would be to just make the initial `kubeadm`-created kubelet client certs have a very short duration like an hour or so.  The kubelet will then auto renew its cert which will be through KCM and thus will honor the configured duration.  @neolit123 can tell me if I have missed anything.

**Comment by neolit123** [comment]
it seems pointless to have a discussion on a closed PR not bound to a design proposal.
@enj i suggest you point the users here to an existing tracking issue for the feature and lock this PR.

to the Qs:

kubeadm signs a kubelet client cert only during bootstrap of the first node and it's for 1 year, after that it's deleted and the client cert produced by signing a CSR (prepared via the kubelet cert manager) by the KCM which owns a CA key. however it will have the same expiration as the initial bootstrap client cert, so 1 year. after that the rotation will trigger and the flag --cluster-signing-duration will be respected. by feeding your custom kubelet.conf to kubeadm it can have a custom duration of say 5 minutes at which point -cluster-signing-duration will come in effect.

kubeadm 1.31 will new v1beta4 API that can control the expiration of the CA and leaf certificate expiration.
so in 1.31 you could set very short expirations for all kubeadm certs (even the bootstrap 1st CP kubelet cert) and they will be respected.

i don't think there are bugs at play here, so there is no need for backports for older versions i'd say.

**Pull request closed without merging**